### PR TITLE
feat: update to vue-virtual-scroller 1.1.2

### LINKED ISSUE #11663: PDF Pages height are not rendering dynamically
## Observed behavior

Issue observed while testing #11145.

When a PDF is rendered, the height of the first page to be rendered is used, and this height is used for all other pages. The problem is if there are PDFs that have pages with different heights, for example, if there are some pages in landscape and others in portrait.

When a landscape page is the first page to be rendered, the portrait pages overflows its height and renders on top of the following page:
image

When a portrait page is the the first page to be rendered, the landscape pages present a big gap between them and the following pages:
image

This issue is because we use a [virtual list for rendering the pages list](https://github.com/learningequality/kolibri/blob/2722f72b6b2758fb0e4721f84551ec08a90b925f/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L87) and render the pdf pages on demand. But this component renders a fixed-size list, which takes the same height for every list item.
 

## Expected behavior

The PDF Renderer list should render a dynamic-size virtual list. We can do this by keeping the `vue-virtual-scroller` library or installing a new library that allows dynamic-size virtual lists. 

## Steps to reproduce

Import the Kolibri QA channel (import token nakav-mafak), and open the "PDF page orientation" resource.

### PR DESCRIPTION
**Pull request by ThEditor**
## Summary

- [x] Updating to vue-virtual-scroller 1.1.2
- [x] Implementing RecycleScroller
- [x] Using DynamicScroller instead of RecycleScroller
- [x] Fix issue with keeping nearby pages loaded (buffer properly?)
- ~[ ] Fit to width dynamically/provide a button~ (Out of scope)

https://github.com/learningequality/kolibri/assets/48005807/d6751e24-0f89-4704-a2be-1279a5b46f82

## References

Closes #11663

Dependency diff:
- [vue-virtual-scroller](https://github.com/Akryum/vue-virtual-scroller)

## Reviewer guidance

Open any pdf document inside kolibri

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by AlexVelezLl** [comment]
Hi @ThEditor. Thanks for your contribution! :tada: 

Could you please retarget the PR and rebase to the `release-v0.16.x` branch. Here is a guide of how you can do it: https://kolibri-dev.readthedocs.io/en/develop/howtos/rebasing_a_pull_request.html.

Also, you can add your name and GH User on the [AUTHORS.md](https://github.com/learningequality/kolibri/blob/release-v0.16.x/AUTHORS.md) file. :tada:

**Comment by ThEditor** [comment]
I think that should have done it. Thanks again @AlexVelezLl !

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: Seems like this `active` and `index` attributes are not defined here. You need to define them in the `template` above:

```Vue
<template #default="{ item, index, active }">
  ...
</template>
```

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: I think we can live without this additional buffer prop for now. Pdf pages are usually full of content, and probably the user will keep viewing the current page for some time, so we dont need to load 2 pages below meanwhile.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: Item.height is undefined here, as the items of the is are the `pdfPages` array, and each item has this structure:
`{
  page: null,
  resolved: false,
  index: i,
}`

**Review by AlexVelezLl — CHANGES_REQUESTED** [review]
Hi! Thank you @ThEditor! Code looks good, but I have encountered some problems:

* The pdf renderer now presents some performance issues since it now loads some more pages and it renders the same page several times.
  * The loading of several pages could be due to the change on the `buffer` prop, and probably it could be easily handled and discussed what can be the optimal value.
  * But what has the biggest impact on performance is that the list renders each page 3 times on the first render and on every new item that is shown while the user scrolls. This leads to jumps and flashes on the document.

  I measured performance of the [page rendered task](https://github.com/ThEditor/kolibri/blob/e6a4da27b2f65da18857db56fc11ac6062d590fe/kolibri/plugins/pdf_viewer/assets/src/views/PdfPage.vue#L136) before and after the update, and after the change there was some situations where it shows some overflow on some pages although this pdf has the same height for all pages:

  Before:

  https://github.com/learningequality/kolibri/assets/51239030/95ec0432-d47f-441e-a0ea-322e759d0fd7

  After:

  https://github.com/learningequality/kolibri/assets/51239030/7c80e363-6e42-4bbe-97f0-848eef7f575f

  It seems it is because vue-scroller internally mount each page Component some additional times to get the item height. But we dont need to mount the pdf page Component to get its height, since its height is defined by the page height, the zoom and margin and it can be calculated without mounting the component.

  We can maybe look at the library to see if we can provide the height of an item without mounting the component multiple times, since this operation is expensive. Another solution that comes to my mind is to try to render like a "centinel node" that can provide to the DynamicScroller the height without doing the expensive tasks that the PdfPage Component does on mount.

* Also, it seems that now the RecycleList component does not expose the `updateVisibleItems` methods, and [this method](https://github.com/ThEditor/kolibri/blob/e6a4da27b2f65da18857db56fc11ac6062d590fe/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L481) fails.

**Comment by ThEditor** [comment]
> But we dont need to mount the pdf page Component to get its height, since its height is defined by the page height, the zoom and margin and it can be calculated without mounting the component

I'm not sure if I understand correctly. We have a pdf document of pages with unknown heights and that's why we try to utilize `DynamicScroller`. If the heights for each page are known but are variable, `RecycleScroller` exposes a variable height mode. Although it also states that that mode is performance heavy for a lot of items.

I also don't understand how we know the height of an item without actually rendering it? Does the page property expose a height field? Aren't zoom, margin set according to the `firstPageHeight`? I don't see how we would be getting the individual page height through that. @AlexVelezLl

**Comment by ThEditor** [comment]
So, I looked into this further. I tried searching for other libraries, tried implementing infinite scrolling on my own, but ended up back to `vue-virtual-scroller`. I have come up with a solution with doesn't have the mounting issue, however, there are some things to note:
- I've modified the original `vue-virtual-scroller`'s source (basically modified `RecycleScroller` to accept size as a function instead of a value)
- I'm using variable-size mode of `RecycleScroller` which is stated as performance heavy in case of large lists.
- I'm pre-computing the heights of the pdf pages, however, I'm not sure if this has a performance impact because I'm not mounting/rendering the pages

I haven't made this into a commit yet because I'm not sure if this is even the right implementation (it works though). I would be interested in a meet, if possible, to discuss this issue, @AlexVelezLl . Thanks again! (also, merry christmas! :D)

**Comment by AlexVelezLl** [comment]
Hi! @ThEditor. Hope you had a merry christmas! and a happy new year :open_hands:.

Ill let there some responses but probably will see deeply on new year:

---

> I also don't understand how we know the height of an item without actually rendering it? Does the page property expose a height field? 

The height of a page can be obtained with  ([here we calculate it](https://github.com/learningequality/kolibri/blob/c372cd05ddd105a7688db9e3698dc21b842ac3e5/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L193)):
```
pageHeight * this.scale + MARGIN
```
pageHeight can be obtained using the `viewport` object [like this](https://github.com/learningequality/kolibri/blob/c372cd05ddd105a7688db9e3698dc21b842ac3e5/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L312).

So, as you can see, we dont need the page to be rendered (`page.render()`) to obtain the height of the node, but just load the page which is a much cheaper operation.

---

> I'm using variable-size mode of RecycleScroller which is stated as performance heavy in case of large lists.

For what I see, I think `variable-size` mode can do the work, we can see the performance implications.

You can push your changes and after the holidays I can review it, and have a meet if needed! :)

---

> I'm pre-computing the heights of the pdf pages, however, I'm not sure if this has a performance impact because I'm not mounting/rendering the pages

We could probably fill all pages with the `firstPage` height and update after when each page loads.

**Comment by ThEditor** [comment]
Oh, that reply sounds exactly like what I implemented in the most recent commit(s). I'm not home right now, I'll push the code as soon as possible. @AlexVelezLl (hope you had an awesome christmas and new year too!)

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue:90]: Sorry to ask this :sweat_smile:, but it seems that the new version of `vue-virtual-scroller` has some differences on the buffer size, and if we dont add a buffer, we can get results like this: 
image

I tried with `:buffer="itemHeight * 2"` and it worked pretty well. I think we can add it here.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: I would not remove this, to have a reference height of the pages. For most cases this height will be the same for all pages.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: Here we are loading all pages of the document at once, at the first pdf load, this could bring performance issues at first load for all pdf documents, including those whose does not have different page heights (which are the majority), so I would not introduce this additional load on performance just to correct the behavior of a minority of documents. I would suggest to calculate this first item size with the height of the first rendered page, and then when we load new pages, we can update the `size` method:

```JS
          for (let i = 0; i < this.totalPages; i++) {
            this.pdfPages.push({
              page: null,
              resolved: false,
              size: () => {
                return this.itemHeight; // height of the first page to be rendered
                // just saw that the height is calculated below this code, but the idea remains the same, set the first page height for all pages at first load
              },
              index: i,
          }
```

And when the pages [are loaded](https://github.com/ThEditor/kolibri/blob/3b6477672cb4b4339f0c7a600aa44d2eec96c3ae/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L383) we can update the function:

```JS
      showPage(pageNum) {
        // Only bother getting it if the pdfPage object is not already cached in the array
        if (pageNum > 0 && pageNum <= this.totalPages && !this.pdfPages[pageNum - 1].resolved) {
          const pageIndex = pageNum - 1;
          this.getPage(pageNum).then(pdfPage => {
            const { height } = pdfPage.getViewport({ scale: 1 });

            // splice so changes are detected
            this.pdfPages.splice(pageIndex, 1, {
              ...this.pdfPages[pageIndex],
              page: pdfPage,
              resolved: true,
              size: () => {
                return height * this.scale + MARGIN;
              },
            });
          });
        }
      },
```
But this approach can lead to "jumps" in the scroll when pages with different height are loaded:

https://github.com/learningequality/kolibri/assets/51239030/41e970a8-b4f4-4796-9d03-c106c1150516

@rtibbles What are your thoughts on this?

File: kolibri/plugins/pdf_viewer/assets/src/views/RecyclableScroller.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/RecyclableScroller.vue:14]: Could you add a comment with the link to the Github URL for the original file (pinned to the specific commit it was copied from if possible)? That will help if we need to update it in the future!

Also, can you flag with code comments where it has been modified from the original and the reason for that modification? [Here is an example](https://github.com/ThEditor/kolibri/blob/3b6477672cb4b4339f0c7a600aa44d2eec96c3ae/kolibri/plugins/pdf_viewer/assets/src/utils/text_layer_builder.js#L21).

**Review by AlexVelezLl — CHANGES_REQUESTED** [review]
In general it is looking amazing! The vue-virtual-scroller update has bring much more smooth scrolling, and it now renders the items in order in the DOM, just let some thoughts on things that can impact the performance.

**Comment by AlexVelezLl** [comment]
Also, there are some falied gh checks: [this](https://github.com/learningequality/kolibri/actions/runs/7359645928/job/20103042994?pr=11671) and [this](https://github.com/learningequality/kolibri/actions/runs/7359645926/job/20103044951?pr=11671). Could you fix them please? For formatting files you can run `yarn run lint-frontend:format` and for running tests you can run `yarn run test`

**Comment by ThEditor** [comment]
I wasn't sure if these changes are "acceptable", and hence, I didn't implement some changes/ideas I had. Now that I know  these are okay, I can proceed. I'll get back to you on these changes as soon as possible.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (ThEditor) [reply kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue:90]: Sure, I'll add that.

File: kolibri/plugins/pdf_viewer/assets/src/views/RecyclableScroller.vue
Comment (ThEditor) [reply kolibri/plugins/pdf_viewer/assets/src/views/RecyclableScroller.vue:14]: I was planning to add a comment with a link to the original file, but I had second-thoughts on the format lol. I'll add that now.

The modification is only one line for now, but there might be a lot of changes if I try to implement one of the ideas I have for lazy-loading page sizes. So, I'll add the "where" of modifications later.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (ThEditor) [reply kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: Since we are rendering the pages one after the other, I think we can modify RecycleScroller further to load the next/previous page's size dynamically, I'll try to give my idea a shot and see if it works. I'll keep you posted.

Comment (AlexVelezLl) [reply kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: Hi @ThEditor! I talked with Richard, and he agreed on allow the slight jump, and we can after open a follow up issue to try to ameliorate it. But for now we can let the size be redefined when the pages load on demand as I suggested above :open_hands:

**Comment by ThEditor** [comment]
I've made the requested changes, this should be ready for review again.

File: kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue
Comment (AlexVelezLl) [inline kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue]: We can delete these comments

**Review by AlexVelezLl — COMMENTED** [review]
Perfect! It is pretty exicitng :tada::tada::tada: . We are almost done! Just one last very little thing from my side is that this node from the internals of the recycle list is taking some space and it shouldn't. We need to avoid it, and also fix a failed check in the code formatting, but everythnig else is looking fine! 

image

Code looks good and some refactor you have done makes the code more readable! Thanks! This PR also solves a bug we had in the pages order in the DOM :open_hands:. I will let the final approve to @radinamatic.

**Comment by ThEditor** [comment]
The space is due to `ResizeObserver` which is a part of [vue-resize](https://github.com/Akryum/vue-resize) (which is used by `vue-virtual-scroller`). The space occurs because of `about:blank` in the [code](https://github.com/Akryum/vue-resize/blob/cc0b8236a866d3d28dfe6e030cd778c30fb2c613/packages/vue-resize/src/components/ResizeObserver.vue#L49C1-L61C6) :
```
    const object = document.createElement('object')
    this._resizeObject = object
    object.setAttribute('aria-hidden', 'true')
    object.setAttribute('tabindex', -1)
    object.onload = this.addResizeHandlers
    object.type = 'text/html'
    if (isIE) {
      this.$el.appendChild(object)
    }
    object.data = 'about:blank'
    if (!isIE) {
      this.$el.appendChild(object)
    }
```
Making the change would require me to also pull in these files into the project. @AlexVelezLl

**Review by pcenov — CHANGES_REQUESTED** [review]
Hi @ThEditor, thank you for your contribution. While testing I noticed the following issues:

1. Scrolling is no longer working when viewing the pdf-s in a mobile browser (tested both with actual device and Chrome's dev tools).

https://github.com/learningequality/kolibri/assets/79847249/3677f5f0-d17b-4e7c-a67d-9bbb7ec43251

2. Sometimes the completion modal is not being displayed after having completed reading the pdf (perhaps that's because of the extra padding at the bottom?).

https://github.com/learningequality/kolibri/assets/79847249/f85c2b75-e7dd-475c-8d9b-7c5ad27dd914

**Comment by ThEditor** [comment]
I am a bit busy for a few days, I'll get back to you as soon as possible.

**Comment by ThEditor** [comment]
Investigated the scrolling issue a bit, broken since [ec1d3a3](https://github.com/learningequality/kolibri/pull/11671/commits/ec1d3a3349eca79c7e813bb2cc11f01c9b2c6cde). Looking into it.

**Comment by ThEditor** [comment]
The scrolling issue on mobile has been fixed. I'll try look into the second issue.

**Comment by ThEditor** [comment]
The extra padding at the bottom has now been fixed, however, I'm not sure how to check if the completion modal issue is also fixed along with it. Also, how do I mark a resource as "not completed"?

**Comment by AlexVelezLl** [comment]
Thanks! @ThEditor I have tested it and I see the extra padding is gone. Waiting for a re-review from @pcenov to confirm everything is ok! Meanwhile if you can rebase the base branch and solve the conflic would be nice :).

**Comment by ThEditor** [comment]
Does this look correct? or did I mess up? 😅 @AlexVelezLl

**Comment by AlexVelezLl** [comment]
Yes! That's exactly how it should look @ThEditor :smiley:. Just noted that a linting check failed hehe, Would be good If you can pass the lint again and update the pr ^^.

**Comment by ThEditor** [comment]
Yeah, I'll try fixing the linting issues. I can't seem to get pre-commit to install locally, hence why code with lint issues.

**Comment by ThEditor** [comment]
Please @ me when you guys have reviewed this, thanks!

**Comment by AlexVelezLl** [comment]
Hello @ThEditor! Thanks for your patience, we will do a final QA review but it won't be merged until 0.16 Kolibri version is released (which should happen in a few weeks). We will keep in touch :open_hands:.

**Review by pcenov — APPROVED** [review]
Thanks @ThEditor - the scrolling issue in mobile devices is fixed now and I didn't observe any regression issues - good to go!

**Review by radinamatic — APPROVED** [review]
Manual QA passed, good to go! 💯 👏🏽 :shipit:

**Review by AlexVelezLl — APPROVED** [review]
Thanks for all your hard work here! @ThEditor :tada: :open_hands:. I've given the code a final check and it looks good! It is ready to go now that 0.16.0 have been released! :tada: :tada: :tada:

**Comment by ThEditor** [comment]
Thanks a lot for merging! Glad to be able to contribute.

**Pull request merged**
